mlrun/feature_store/steps.py

```diff
@@ -226,15 +227,68 @@ def _do_pandas(self, event):
     def _do_spark(self, event):
         from itertools import chain
 
-        from pyspark.sql.functions import col, create_map, lit, when
+        from pyspark.sql.functions import col, create_map, isnan, isnull, lit, when
+        from pyspark.sql.types import DecimalType, DoubleType, FloatType
+        from pyspark.sql.utils import AnalysisException
 
         for column, column_map in self.mapping.items():
             new_column_name = self._get_feature_name(column)
             if "ranges" not in column_map:
                 mapping_expr = create_map([lit(x) for x in chain(*column_map.items())])
-                event = event.withColumn(
-                    new_column_name, mapping_expr.getItem(col(column))
-                )
+                try:
+                    event = event.withColumn(
+                        new_column_name,
+                        when(
+                            col(column).isin(list(column_map.keys())),
+                            mapping_expr.getItem(col(column)),
+                        ).otherwise(col(column)),
+                    )
+                #  if failed to use otherwise it is probably because the new column has different type
+                #  then the original column.
+                #  we will try to replace the values without using 'otherwise'.
+                except AnalysisException:
+                    event = event.withColumn(
+                        new_column_name, mapping_expr.getItem(col(column))
+                    )
+                    col_type = event.schema[column].dataType
+                    new_col_type = event.schema[new_column_name].dataType
+                    #  in order to avoid exception at isna on non-decimal/float columns -
+                    #  we need to check their types before filtering.
+                    if isinstance(col_type, (FloatType, DoubleType, DecimalType)):
+                        column_filter = (~isnull(col(column))) & (~isnan(col(column)))
+                    else:
+                        column_filter = ~isnull(col(column))
+                    if isinstance(new_col_type, (FloatType, DoubleType, DecimalType)):
+                        new_column_filter = isnull(col(new_column_name)) | isnan(
+                            col(new_column_name)
+                        )
+                    else:
+                        #  we need to check that every value replaced if we changed column type - except None or Nan.
+                        new_column_filter = isnull(col(new_column_name))
+                    turned_to_none_values = (
+                        event.filter(column_filter & new_column_filter)
+                        .select(column)
+                        .rdd.map(lambda x: x[0])
+                        .collect()
```

Comment:
I don't think you can/should collect the entire dataset to memory. The whole point of spark is that you shouldn't have to.